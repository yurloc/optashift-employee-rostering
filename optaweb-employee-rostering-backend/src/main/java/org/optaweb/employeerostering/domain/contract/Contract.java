/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.optaweb.employeerostering.domain.contract;

import javax.persistence.Entity;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;
import javax.validation.constraints.AssertTrue;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import javax.validation.constraints.Size;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.optaweb.employeerostering.domain.common.AbstractPersistable;

@Entity
@Table(uniqueConstraints = {@UniqueConstraint(columnNames = {"tenantId", "name"}),
        @UniqueConstraint(columnNames = {"id"})})
// TODO: Single Responsibility Principle - acts as both domain entity and JSON-serializable entity
public class Contract extends AbstractPersistable {

    @NotNull
    @Size(min = 1, max = 120)
    @Pattern(regexp = "^(?!\\s).*(?<!\\s)$", message = "Name should not contain any leading or trailing whitespaces")
    private String name;

    // Can be null
    @Min(0)
    @Max(60 * 24)
    private Integer maximumMinutesPerDay;

    // Can be null
    @Min(0)
    @Max(60 * 24 * 7)
    private Integer maximumMinutesPerWeek;

    // Can be null
    // 31 days in longest month
    @Min(0)
    @Max(60 * 24 * 31)
    private Integer maximumMinutesPerMonth;

    // Can be null
    // 366 days in leap year
    @Min(0)
    @Max(60 * 24 * 366)
    private Integer maximumMinutesPerYear;

    @SuppressWarnings("unused")
    public Contract() {
    }

    public Contract(Integer tenantId, String name) {
        this(tenantId, name, null, null, null, null);
    }

    public Contract(Integer tenantId, String name, Integer maximumMinutesPerDay, Integer maximumMinutesPerWeek,
                    Integer maximumMinutesPerMonth, Integer maximumMinutesPerYear) {

        super(tenantId);
        this.name = name;
        this.maximumMinutesPerDay = maximumMinutesPerDay;
        this.maximumMinutesPerWeek = maximumMinutesPerWeek;
        this.maximumMinutesPerMonth = maximumMinutesPerMonth;
        this.maximumMinutesPerYear = maximumMinutesPerYear;
    }

    @AssertTrue
    @JsonIgnore
    public boolean isValid() {
        if (maximumMinutesPerDay != null && maximumMinutesPerDay <= 0) {
            return false;
        }
        if (maximumMinutesPerWeek != null && maximumMinutesPerWeek <= 0) {
            return false;
        }
        if (maximumMinutesPerMonth != null && maximumMinutesPerMonth <= 0) {
            return false;
        }
        if (maximumMinutesPerYear != null && maximumMinutesPerYear <= 0) {
            return false;
        }
        return true;
    }

    @Override
    public String toString() {
        return name;
    }

    // ************************************************************************
    // Simple getters and setters
    // ************************************************************************

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Integer getMaximumMinutesPerDay() {
        return maximumMinutesPerDay;
    }

    public void setMaximumMinutesPerDay(Integer maximumMinutesPerDay) {
        this.maximumMinutesPerDay = maximumMinutesPerDay;
    }

    public Integer getMaximumMinutesPerWeek() {
        return maximumMinutesPerWeek;
    }

    public void setMaximumMinutesPerWeek(Integer maximumMinutesPerWeek) {
        this.maximumMinutesPerWeek = maximumMinutesPerWeek;
    }

    public Integer getMaximumMinutesPerMonth() {
        return maximumMinutesPerMonth;
    }

    public void setMaximumMinutesPerMonth(Integer maximumMinutesPerMonth) {
        this.maximumMinutesPerMonth = maximumMinutesPerMonth;
    }

    public Integer getMaximumMinutesPerYear() {
        return maximumMinutesPerYear;
    }

    public void setMaximumMinutesPerYear(Integer maximumMinutesPerYear) {
        this.maximumMinutesPerYear = maximumMinutesPerYear;
    }
}
