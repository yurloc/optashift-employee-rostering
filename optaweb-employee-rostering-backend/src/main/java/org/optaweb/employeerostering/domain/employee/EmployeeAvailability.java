/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.optaweb.employeerostering.domain.employee;

import java.time.Duration;
import java.time.OffsetDateTime;
import java.time.ZoneId;

import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.ManyToOne;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;
import javax.validation.constraints.AssertTrue;
import javax.validation.constraints.NotNull;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.optaweb.employeerostering.domain.common.AbstractPersistable;
import org.optaweb.employeerostering.domain.employee.view.EmployeeAvailabilityView;

@Entity
@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"tenantId", "employee_id", "startDateTime", "endDateTime"}))
// TODO: Single Responsibility Principle - acts as both domain entity and JSON-serializable entity
public class EmployeeAvailability extends AbstractPersistable {

    @NotNull
    @ManyToOne(fetch = FetchType.EAGER)
    private Employee employee;

    @NotNull
    private OffsetDateTime startDateTime;
    @NotNull
    private OffsetDateTime endDateTime;

    @NotNull
    private EmployeeAvailabilityState state;

    @SuppressWarnings("unused")
    public EmployeeAvailability() {
    }

    public EmployeeAvailability(Integer tenantId, Employee employee, OffsetDateTime startDateTime,
                                OffsetDateTime endDateTime) {
        super(tenantId);
        this.employee = employee;
        this.startDateTime = startDateTime;
        this.endDateTime = endDateTime;
    }

    public EmployeeAvailability(ZoneId zoneId, EmployeeAvailabilityView employeeAvailabilityView, Employee employee) {
        super(employeeAvailabilityView);
        this.employee = employee;
        this.startDateTime = OffsetDateTime.of(employeeAvailabilityView.getStartDateTime(),
                                             zoneId.getRules().getOffset(employeeAvailabilityView.getStartDateTime()));
        this.endDateTime = OffsetDateTime.of(employeeAvailabilityView.getEndDateTime(),
                                             zoneId.getRules().getOffset(employeeAvailabilityView.getEndDateTime()));
        this.state = employeeAvailabilityView.getState();
    }

    @AssertTrue
    @JsonIgnore
    public boolean isValid() {
        return getDuration().getSeconds() / (60 * 60) < 28;
    }

    @JsonIgnore
    public Duration getDuration() {
        return Duration.between(startDateTime, endDateTime);
    }

    @Override
    public String toString() {
        return employee + ":" + startDateTime + "-" + endDateTime;
    }

    // ************************************************************************
    // Simple getters and setters
    // ************************************************************************

    public Employee getEmployee() {
        return employee;
    }

    public void setEmployee(Employee employee) {
        this.employee = employee;
    }

    public OffsetDateTime getStartDateTime() {
        return startDateTime;
    }

    public void setStartDateTime(OffsetDateTime startDateTime) {
        this.startDateTime = startDateTime;
    }

    public OffsetDateTime getEndDateTime() {
        return endDateTime;
    }

    public void setEndDateTime(OffsetDateTime endDateTime) {
        this.endDateTime = endDateTime;
    }

    public EmployeeAvailabilityState getState() {
        return state;
    }

    public void setState(EmployeeAvailabilityState state) {
        this.state = state;
    }

    public EmployeeAvailability inTimeZone(ZoneId zoneId) {
        return new EmployeeAvailability(zoneId, new EmployeeAvailabilityView(zoneId, this), getEmployee());
    }
}
