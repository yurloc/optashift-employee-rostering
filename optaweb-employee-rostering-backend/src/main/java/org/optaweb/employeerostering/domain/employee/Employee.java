/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.optaweb.employeerostering.domain.employee;

import java.util.Collection;
import java.util.Set;

import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import javax.validation.constraints.Size;

import org.optaweb.employeerostering.domain.common.AbstractPersistable;
import org.optaweb.employeerostering.domain.contract.Contract;
import org.optaweb.employeerostering.domain.skill.Skill;

@Entity
@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"tenantId", "name"}))
public class Employee extends AbstractPersistable {

    @NotNull
    @Size(min = 1, max = 120)
    @Pattern(regexp = "^(?!\\s).*(?<!\\s)$", message = "Name should not contain any leading or trailing whitespaces")
    private String name;

    @NotNull
    @ManyToOne
    private Contract contract;

    @NotNull
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(name = "EmployeeSkillProficiencySet",
            joinColumns = @JoinColumn(name = "employeeId", referencedColumnName = "id"),
            inverseJoinColumns = @JoinColumn(name = "skillId", referencedColumnName = "id")
    )
    private Set<Skill> skillProficiencySet;

    @SuppressWarnings("unused")
    public Employee() {
    }

    public Employee(Integer tenantId, String name, Contract contract, Set<Skill> skillProficiencySet) {
        super(tenantId);
        this.name = name;
        this.contract = contract;
        this.skillProficiencySet = skillProficiencySet;
    }

    public boolean hasSkill(Skill skill) {
        return skillProficiencySet.contains(skill);
    }

    public boolean hasSkills(Collection<Skill> skills) {
        return skillProficiencySet.containsAll(skills);
    }

    @Override
    public String toString() {
        return name;
    }

    // ************************************************************************
    // Simple getters and setters
    // ************************************************************************

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Set<Skill> getSkillProficiencySet() {
        return skillProficiencySet;
    }

    public void setSkillProficiencySet(Set<Skill> skillProficiencySet) {
        this.skillProficiencySet = skillProficiencySet;
    }

    public Contract getContract() {
        return contract;
    }

    public void setContract(Contract contract) {
        this.contract = contract;
    }

    @Override
    public boolean equals(Object o) {
        if (o instanceof Employee) {
            Employee other = (Employee) o;
            return this.name.equals(other.getName()) &&
                    this.contract.equals(other.getContract()) &&
                    this.skillProficiencySet.equals(other.getSkillProficiencySet());
        } else {
            return false;
        }
    }

    @Override
    public int hashCode() {
        return 31 * ((31 * name.hashCode()) ^ contract.hashCode()) ^ skillProficiencySet.hashCode();
    }
}
