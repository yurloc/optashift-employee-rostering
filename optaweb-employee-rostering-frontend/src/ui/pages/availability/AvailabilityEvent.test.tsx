/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Contract } from 'domain/Contract';
import { Employee } from 'domain/Employee';
import { EmployeeAvailability } from 'domain/EmployeeAvailability';
import moment from 'moment';
import AvailabilityEvent, { AvailabilityEventProps } from './AvailabilityEvent';

describe('Availability Event', () => {
  it('should render correctly', () => {
    const event = shallow(<AvailabilityEvent {...baseProps} />);
    expect(toJson(event)).toMatchSnapshot();
  });

  it('should call removeEmployeeAvailability if they click on the delete button', () => {
    const event = shallow(<AvailabilityEvent {...baseProps} />);
    event.find('[aria-label="Delete"]').simulate('click');
    expect(baseProps.removeEmployeeAvailability).toBeCalled();
    expect(baseProps.removeEmployeeAvailability).toBeCalledWith(availability);
  });

  it('should call updateEmployeeAvailability with a "DESIRED" state'
    + 'if they click on the "DESIRED" button', () => {
    const event = shallow(<AvailabilityEvent
      {...baseProps}
      availability={{
        ...baseProps.availability,
        state: 'UNAVAILABLE',
      }}
    />);
    event.find('[aria-label="Desired"]').simulate('click');
    expect(baseProps.updateEmployeeAvailability).toBeCalled();
    expect(baseProps.updateEmployeeAvailability).toBeCalledWith(availability);
  });

  it('should call updateEmployeeAvailability with a "UNDESIRED" state'
  + 'if they click on the "UNDESIRED" button', () => {
    const event = shallow(<AvailabilityEvent {...baseProps} />);
    event.find('[aria-label="Undesired"]').simulate('click');
    expect(baseProps.updateEmployeeAvailability).toBeCalled();
    expect(baseProps.updateEmployeeAvailability).toBeCalledWith({
      ...baseProps.availability,
      state: 'UNDESIRED',
    });
  });

  it('should call updateEmployeeAvailability with a "UNAVAILABLE" state'
  + 'if they click on the "UNAVAILABLE" button', () => {
    const event = shallow(<AvailabilityEvent {...baseProps} />);
    event.find('[aria-label="Unavailable"]').simulate('click');
    expect(baseProps.updateEmployeeAvailability).toBeCalled();
    expect(baseProps.updateEmployeeAvailability).toBeCalledWith({
      ...baseProps.availability,
      state: 'UNAVAILABLE',
    });
  });
});

const contract: Contract = {
  id: 0,
  version: 0,
  tenantId: 0,
  name: 'Contract',
  maximumMinutesPerDay: 100,
  maximumMinutesPerWeek: 1000,
  maximumMinutesPerMonth: 10000,
  maximumMinutesPerYear: 100000,
};

const employee: Employee = {
  id: 1,
  version: 0,
  tenantId: 0,
  name: 'Amy',
  contract,
  skillProficiencySet: [],
};

const startDateTime = moment('2018-01-01T09:00').toDate();
const endDateTime = moment('2018-01-01T17:00').toDate();

const availability: EmployeeAvailability = {
  id: 2,
  version: 0,
  tenantId: 0,
  employee,
  startDateTime,
  endDateTime,
  state: 'DESIRED',
};

const baseProps: AvailabilityEventProps = {
  availability,
  onEdit: jest.fn(),
  onDelete: jest.fn(),
  updateEmployeeAvailability: jest.fn(),
  removeEmployeeAvailability: jest.fn(),
};
