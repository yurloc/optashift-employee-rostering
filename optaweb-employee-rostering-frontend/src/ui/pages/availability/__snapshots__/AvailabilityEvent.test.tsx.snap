// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Availability Event should render correctly 1`] = `
<span
  className="availability-event"
>
  <Split>
    <SplitItem
      isFilled={false}
    >
      Trans(i18nKey=EmployeeAvailabilityState.DESIRED)
    </SplitItem>
    <SplitItem
      isFilled={true}
    />
    <SplitItem
      isFilled={false}
    >
      <Button
        aria-label="Delete"
        onClick={[Function]}
        variant="danger"
      >
        <TrashIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          title={null}
        />
      </Button>
    </SplitItem>
  </Split>
  <Level
    gutter="sm"
  >
    <LevelItem>
      <Button
        aria-label="Desired"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "green",
            "margin": "1px",
            "width": "min-content",
          }
        }
        title="Trans(i18nKey=EmployeeAvailabilityState.DESIRED)"
        variant="tertiary"
      >
        <OkIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          title={null}
        />
      </Button>
      <Button
        aria-label="Undesired"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "yellow",
            "margin": "1px",
            "width": "min-content",
          }
        }
        title="Trans(i18nKey=EmployeeAvailabilityState.UNDESIRED)"
        variant="tertiary"
      >
        <WarningTriangleIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          title={null}
        />
      </Button>
      <Button
        aria-label="Unavailable"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "red",
            "margin": "1px",
            "width": "min-content",
          }
        }
        title="Trans(i18nKey=EmployeeAvailabilityState.UNAVAILABLE)"
        variant="tertiary"
      >
        <ErrorCircleOIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          title={null}
        />
      </Button>
    </LevelItem>
  </Level>
</span>
`;
